libs/log/sem_logger.go

```diff
@@ -0,0 +1,176 @@
+
+const (
+	SemTransaction SemEntity = iota + 1
+	SemAddress
+)
+
+type SemStatus struct {
+	Active bool
+	Rules  map[SemEntity]map[string]bool
+}
+
+type semLogger struct {
+	logBackend Logger
+	rules      map[SemEntity]map[string]uint
+	active     bool
+	enabled    bool
+}
+
+func (l *semLogger) Error(msg string, kvals ...interface{}) {
+	l.logBackend.Error(msg, kvals...)
+}
+
+func (l *semLogger) Info(msg string, kvals ...interface{}) {
+	if !l.enabled || l.active {
+		l.logBackend.Info(msg, kvals...)
+	}
+}
+
+func (l *semLogger) Debug(msg string, kvals ...interface{}) {
+	if !l.enabled || l.active {
+		l.logBackend.Debug(msg, kvals...)
+	}
+}
+
+func (l *semLogger) With(keyvals ...interface{}) Logger {
+	sl := new(semLogger)
+	sl.logBackend = l.logBackend.With(keyvals...)
+	sl.rules = l.rules
+	sl.setActiveState(l.active)
+	return sl
+}
+
+func (l *semLogger) AddRule(entity SemEntity, value []byte) {
+	_, exist := l.rules[entity]
+	key := string(value)
+	if !exist {
+		l.rules[entity] = map[string]uint{key: 0}
+	} else {
+		_, exist = l.rules[entity][key]
+		if !exist {
+			l.rules[entity][key] = 0
+		}
+	}
+	l.setEnabled(true)
+}
+
+func (l *semLogger) DeleteAll() {
+	for rule := range l.rules {
+		delete(l.rules, rule)
+	}
+	l.setEnabled(false)
+	l.setActiveState(false)
+}
+
+func (l *semLogger) Status() (status SemStatus, err error) {
+	status = SemStatus{
+		Active: l.active,
+		Rules:  map[SemEntity]map[string]bool{},
+	}
+	for entity, rules := range l.rules {
+		for rule, count := range rules {
+			_, exists := status.Rules[entity]
+			if !exists {
+				status.Rules[entity] = map[string]bool{}
+			}
+			status.Rules[entity][rule] = count > 0
+
+		}
+	}
+	return
+}
+
+// Set SEM logger state to active
+// In active state all log levels are enabled.
+// In deactivated state (no matching rule) only error logs are forwarded to logger backend
+func (l *semLogger) setActiveState(active bool) {
+	// For Filter loggers we need to make sure that no filtering is happening underneath
+	next, isFilterLogger := l.logBackend.(*filter)
+	if isFilterLogger {
+		next.SetSemStatus(active)
+	}
+	l.active = active
+}
+
+// Disable/Enable SEM logger
+// In disabled state (no rules defined) no 'filterring' is done on SEM level and no bypass
+// on filter-capable backends should be done
+func (l *semLogger) setEnabled(enable bool) {
+	l.enabled = enable
+	// For Filter loggers we need to make sure that no filtering is happening underneath
+	next, isFilterLogger := l.logBackend.(*filter)
+	if isFilterLogger {
+		next.SetSemStatus(enable)
+	}
+}
+
+func (l *semLogger) Entry(entity SemEntity, value []byte) {
+	_, exist := l.rules[entity][string(value)]
+	if !exist {
+		return
+	}
+	if !l.active {
+		l.setActiveState(true)
+	}
+	l.rules[entity][string(value)]++
+}
+
+func (l *semLogger) Exit(entity SemEntity, value []byte) {
+	_, exist := l.rules[entity][string(value)]
+	if !exist {
+		return
+	}
+
+	l.rules[entity][string(value)]-- //TODO log on underflow
```

Comment:
Apart from the error log, shouldn't we also keep the counter to 0 (effectively turning the "excess" `Exit` into a noop?
TBH, not sure what we should do... something think about